src/sc2api/sc2_agent.cc

```diff
@@ -122,65 +128,131 @@ void ActionImp::SendChat(const std::string& message, ChatChannel channel) {
 
 void ActionImp::UnitCommand(const Unit* unit, AbilityID ability, bool queued_command) {
     if (!unit) return;
-    UnitCommand(Units({ unit }), ability, queued_command);
+    Tag tag = unit->tag;
+    UnitCommand(tag, ability, queued_command);
 }
 
 void ActionImp::UnitCommand(const Unit* unit, AbilityID ability, const Point2D& point, bool queued_command) {
     if (!unit) return;
-    UnitCommand(Units({ unit }), ability, point, queued_command);
+    Tag tag = unit->tag;
+    UnitCommand(tag, ability, point, queued_command);
 }
 
 void ActionImp::UnitCommand(const Unit* unit, AbilityID ability, const Unit* target, bool queued_command) {
     if (!unit || !target) return;
-    UnitCommand(Units({ unit }), ability, target, queued_command);
+    Tag tag = unit->tag;
+    Tag target_tag = target->tag;
+    UnitCommand(tag, ability, target_tag, queued_command);
 }
 
 void ActionImp::UnitCommand(const Units& units, AbilityID ability, bool queued_command) {
+    std::vector<Tag> tags;
+    for (auto unit : units) {
+        if (!unit) continue;
+        tags.push_back(unit->tag);
+        UnitCommand(tags, ability, queued_command);
+    }
```

Comment:
It seems to me that `std::copy_if` is unable to insert `unit->tag` during it's operation?
https://en.cppreference.com/w/cpp/algorithm/copy
> The expression pred(v) ... must not modify v.

At first I was trying to std::copy_if(units.begin(), units.end(), std::back_inserter(tags), is_valid());
And in the predicate is_valid(), I was playing with lambdas and after yelling at my compiler, realized my error.

At the moment, I am doing some review of lambdas and passing a struct to output a data member.  I'll give it another look today and see if I can more accurately define my conceptual deficiencies.

Reply:
Ah, I forgot that we need to put Tag, not Unit. Well, in this case it is ok to use current approach.
It is possible to do such action by combining filter and transform (also boost provides interesting capabilities), but this solution is relatively complex for such simple task.

Reply:
It was good practice trying to apply std library, since I have little experience with it, and it forced me to start attempting to use reference materials.